README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## megalearner

<!-- badges: start -->

<!-- badges: end -->

**megalearner** is series of R scripts, found in the folder R, that
implement the illustrative example described in the paper Gonzalez
Ginestet, P. et al. (2021). “Survival stacking with multiple data types
using pseudo-observation-based-AUC loss”.

**megalearner** builds a more accurate risk prediction model than
standard methods such as Cox proportional hazard or single machine
learning algorithms that have been already adapted to handle
right-censored observations. **megalearner** optimally stacks
predictions from any method based on different survival outcome types
(either binary or continuous). In the paper, we show that our proposed
method can improve on single survival based methods such as survival
random forest or on other strategies that use a pre-processing step such
as inverse probability of censoring weighted or pseudo-observations.

We use the Rotterdam data set as training data set and we validate the
predictions in an external data set. For that, we use the data from the
German Breast Cancer Study Group (GBSG). Both data sets are available in
the `survival` R package \[Therneau, 2000\].

## Organization of the folders

The scripts depend on each other in the following way:

03-validate.R -\> depends on -\> 02-opt-stack.R -\> depends on -\>
01-train-models.R -\> depends on 00-model-specs.R

01-train-models.R: saves the fit of each algorithm (“fullfits.rds”), the
matrix Z (“Zout.rds”) where the predictions of each algorithms are
recorded and the folds used in the cross-validation (“split.rds”)

02-opt-stack.R: saves the optimal coefficient used to combine the
algorithms (“opt-coeffs.rds”) and the AUC at 5 years cross-validated of
the stacking (“cv-aucs-stack.rds”)

The folder figure contains the plots in the paper (Figure 1 and Figure
2).

## Installation

Depends on the following R packages: `broom`, `patchwork`, `ggplot2`,
`xtable`, `class`, `glmnet`, `e1071`, `riskRegression`, `CoxBoost`,
`prodlim`, `Matrix`, `randomForestSRC`, `survival`, `eventglm`,
`pseudoloss`, `splines`, `stats`, `graphics`, `grDevices`, `utils`,
`datasets`, `methods`, `base`, `plotROC`

The following packages `CoxBoost`, `eventglm` and `pseudoloss` can be
installed as follows:

``` r
remotes::install_github("binderh/CoxBoost")
remotes::install_github("sachsmc/eventglm")
remotes::install_github("sachsmc/pseudoloss")
```


NEWS.md
1
# megalearner 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/00-model-specs.R
library(eventglm) ## remotes::install_github("sachsmc/eventglm")
library(survival)
library(randomForestSRC)
library(CoxBoost)  ## remotes::install_github("binderh/CoxBoost")
library(riskRegression)
library(e1071)
library(splines)
library(glmnet)
library(class)

## models

## survival
############### Natively ############################
#' @description  Predictions based on those Machine Learning procedures in the library that allow for weights to be specified as an argument of the R function. No bagging occurs. This group of algorithms is denoted as  Native Weights
#' @param dset data set
#' @return  prediction
#' @rdname megalearner-internal
#'
stepwise <- function(dset) {

  cfit <- coxph(YY ~ ., data = dset)
  cfin <- step(cfit, trace = 0)
  function(valid) {
    predict(cfin, newdata = valid,
            type = "lp")
  }

}

random.forest <- function(dset, time = 5 * 365.25) {

  dset$time <- dset$YY[, "time"]
  dset$status <- dset$YY[, "status"]
  dset$YY <-  NULL

  cfit <- rfsrc(Surv(time, status) ~ ., data = dset)
  function(valid) {
    res <- predict(cfit, newdata = valid[, -1])
    1 - res$survival[, max(which(res$time.interest <= time))[1]]
  }

}

coxboost <- function(dset) {

  mm <- model.matrix(YY ~ ., data = dset)[, -1]
  cfit <- CoxBoost(time = dset$YY[, "time"], status = dset$YY[, "status"],
           x = mm)
  function(valid) {
    mmv <- model.matrix(YY ~ ., data = valid)[, -1]
    res <- predict(cfit, newdata = mmv,
                   type = "lp")
    res[1, ]
  }

}

## binary

direct.binomial <- function(dset, time = 5 * 365.25) {
  dset$time <- dset$YY[, "time"]
  dset$status <- dset$YY[, "status"]
  dset$YY <-  NULL
  dset$ybin <- 1.0 * (dset$time < time & dset$status == 1)
  dset$ybin[dset$time < time & dset$status == 0] <- NA

  sfit <- survfit(Surv(time, 1 - status) ~ 1, data = dset)
  dset$weights <- 1 / summary(sfit, times = pmin(dset$time, time))$surv
  dset$time <- dset$status <- NULL

  cfit <- glm(ybin ~ bs(age) + meno + sizemed + sizebig + bs(nodes) + bs(pgr) + bs(er) +
                hormon,
      data = dset, family = "binomial", weights = weights)
  cfin <- step(cfit, trace = 0)
  function(valid) {
      predict(cfit, newdata = valid, type = "response")
  }

}

## bagging ipcw

svm <- function(dset, time = 5 * 365.25){

  dset$time <- dset$YY[, "time"]
  dset$status <- dset$YY[, "status"]
  dset$YY <-  NULL
  dset$ybin <- 1.0 * (dset$time < time & dset$status == 1)
  dset$ybin[dset$time < time & dset$status == 0] <- NA

  sfit <- survfit(Surv(time, 1 - status) ~ 1, data = dset)

  dset <- dset[!is.na(dset$ybin),]
  wtmp <- 1 / summary(sfit, times = pmin(dset$time, time))$surv
  dset$samp.wts <- wtmp / sum(wtmp)
  dset$time <- dset$status <- NULL

  svmboot <- lapply(1:10, function(i) {

    dboot <- dset[sample(1:nrow(dset), nrow(dset),
                         replace = TRUE, prob = dset$samp.wts),]
    dboot$samp.wts <- NULL
    e1071::svm(ybin ~ ., data = dboot)

  })


  function(valid) {

    vboots <- sapply(svmboot, function(cfit) {
      predict(cfit, newdata = valid)
    })
    rowMeans(vboots)

  }
}



knn <- function(dset, time = 5 * 365.25){

  dset$time <- dset$YY[, "time"]
  dset$status <- dset$YY[, "status"]
  dset$YY <-  NULL
  dset$ybin <- 1.0 * (dset$time < time & dset$status == 1)
  dset$ybin[dset$time < time & dset$status == 0] <- NA

  sfit <- survfit(Surv(time, 1 - status) ~ 1, data = dset)

  dset <- dset[!is.na(dset$ybin),]
  wtmp <- 1 / summary(sfit, times = pmin(dset$time, time))$surv
  dset$samp.wts <- wtmp / sum(wtmp)
  dset$time <- dset$status <- NULL


  function(valid) {

    vboots <- sapply(1:10, function(i) {

      dboot <- dset[sample(1:nrow(dset), nrow(dset),
                           replace = TRUE, prob = dset$samp.wts),]
      dboot$samp.wts <- NULL
      tclass <- as.factor(dboot$ybin)
      dboot$ybin <- NULL

      knboot <- class::knn(dboot, valid[, colnames(dboot)],
          cl = tclass, k = 5, prob = TRUE)

      ifelse(knboot == 0, 1 - attr(knboot, "prob"), attr(knboot, "prob"))

    })

    rowMeans(vboots)

  }
}


## pseudo obs

pseudo.glm <- function(dset, time = 5 * 365.25){

  cfit <- cumincglm(YY ~ bs(age) + meno + sizemed + sizebig + bs(nodes) + bs(pgr) + bs(er) +
                      hormon, data = dset, time = time)
  function(valid) {
    predict(cfit, newdata = valid, type = "response")
  }

}


pseudo.glmnet <- function(dset, time = 5 * 365.25) {

  cfit <- cumincglm(YY ~ bs(age) + meno + sizemed + sizebig + bs(nodes) + bs(pgr) + bs(er) +
                      hormon, data = dset, time = time, x = TRUE)

  cnfit <- cv.glmnet(cfit$x, cfit$y)

  function(valid) {
    vx <- model.matrix(~ bs(age) + meno + sizemed + sizebig + bs(nodes) + bs(pgr) + bs(er) +
                         hormon, data = valid)
    predict(cnfit, newx = vx)[, 1]

  }

}




R/02-opt-stack.R
library(pseudoloss) ## remotes::install_github("sachsmc/pseudoloss")
#source("00-model-specs.R")

Zout <- readRDS("Zout.rds")
fullfits <- readRDS("fullfits.rds")
folds <- readRDS("split.rds")

data("gbsg")
data("rotterdam")

rotterdam <- subset(rotterdam, nodes > 0)

devel <- data.frame(YY = with(rotterdam, Surv(pmin(rtime, dtime), 1.0 * (recur == 1 | unclass(death) == 1))),
                    age = rotterdam$age, meno = unclass(rotterdam$meno),
                    sizemed = 1.0 * (rotterdam$size == "20-50"),
                    sizebig = 1.0 * (rotterdam$size == ">50"),
                    nodes = unclass(rotterdam$nodes), pgr = unclass(rotterdam$pgr),
                    er = unclass(rotterdam$er), hormon = unclass(rotterdam$hormon))

devel$PO <- cumincglm(YY ~ 1, data = devel, time = 5 * 365.25)$y

YYens <- devel$PO[unlist(folds)]
Zmat <- do.call(rbind, Zout)

lassofit <- cv.glmnet(Zmat, YYens, standardize = FALSE, alpha = 0)
lfit <- glmnet(Zmat, YYens, standardize = FALSE,
               lambda = lassofit$lambda.1se, alpha = 0)

kcand <- exp(seq(-5, 0))

set.seed(1890)
ndex <- 1:nrow(Zmat)
part <- as.factor(sample(1:10, length(ndex), replace = TRUE))
folds <- split(ndex, part)

cv.auc <- matrix(NA, nrow = 10, ncol = length(kcand))

for(k in 1:length(kcand)) {
  for(i in 1:length(folds)){

    Zin <- Zmat[unlist(folds[-i]),]
    Zout <- Zmat[folds[[i]],]
    YYin <- YYens[unlist(folds[-i])]
    YYout <- YYens[folds[[i]]]

    fn.auc <- function(beta) {

      yy <- c(Zin %*% matrix(beta, ncol = 1))
      1 - with(calc_roc(yy, matrix(YYin, ncol = 1), ramp = smoothramp),
               calc_auc(fpf, tpf)) + k * sum(abs(beta)^2)
    }

    start.beta <- as.matrix(lfit$beta)[, 1]
    opt.fit <- optim(par = start.beta, fn = fn.auc, method = "BFGS")

    cv.auc[i, k] <- with(calc_roc((Zout %*% opt.fit$par)[, 1],
                                  matrix(YYout, ncol = 1), smoothramp),
                      calc_auc(fpf, tpf))

  }
}

saveRDS(cv.auc, "cv-aucs-stack.rds")

k <- kcand[which.max(colMeans(cv.auc))]
fn.auc <- function(beta) {

  yy <- c(Zmat %*% matrix(beta, ncol = 1))
  1 - with(calc_roc(yy, matrix(YYens, ncol = 1), ramp = smoothramp),
           calc_auc(fpf, tpf)) + k * sum(abs(beta)^2)
}

start.beta <- opt.fit$par

opt.fit <- optim(par = start.beta, fn = fn.auc, method = "BFGS")

saveRDS(opt.fit, "opt-coeffs.rds")


R/01-train-models.R
#source("00-model-specs.R")
data("gbsg")
data("rotterdam")

mymods <- list(stepwise, random.forest, coxboost,
               direct.binomial, svm, knn,
               function(xx) pseudo.glm(xx, time = 2 * 365.25),
               function(xx) pseudo.glm(xx, time = 4 * 365.25),
               function(xx) pseudo.glm(xx, time = 5 * 365.25),
               function(xx) pseudo.glm(xx, time = 6 * 365.25),
               function(xx) pseudo.glm(xx, time = 7 * 365.25),
               function(xx) pseudo.glmnet(xx, time = 2 * 365.25),
               function(xx) pseudo.glmnet(xx, time = 4 * 365.25),
               function(xx) pseudo.glmnet(xx, time = 5 * 365.25),
               function(xx) pseudo.glmnet(xx, time = 6 * 365.25),
               function(xx) pseudo.glmnet(xx, time = 7 * 365.25))

rotterdam <- subset(rotterdam, nodes > 0)

devel <- data.frame(YY = with(rotterdam, Surv(pmin(rtime, dtime), 1.0 * (recur == 1 | unclass(death) == 1))),
                    age = rotterdam$age, meno = unclass(rotterdam$meno),
                    sizemed = 1.0 * (rotterdam$size == "20-50"),
                    sizebig = 1.0 * (rotterdam$size == ">50"),
                    nodes = unclass(rotterdam$nodes), pgr = unclass(rotterdam$pgr),
                    er = unclass(rotterdam$er), hormon = unclass(rotterdam$hormon))

set.seed(420)
ndex <- 1:nrow(devel)
part <- as.factor(sample(1:10, length(ndex), replace = TRUE))
folds <- split(ndex, part)

Zout <- vector(mode = "list", length = 10)
for(j in 1:length(folds)) {

  training <- devel[unlist(folds[-j]), ]
  validation <- devel[folds[[j]], ]

  Zout[[j]] <- do.call(cbind, lapply(mymods, function(f){

    fhat <- f(training)
    fhat(validation)

  }))

}


fullfits <- lapply(mymods, function(f) {

  f(devel)

})


saveRDS(fullfits, "fullfits.rds")
saveRDS(Zout, "Zout.rds")
saveRDS(folds, "split.rds")


R/03-validate.R
library(pseudoloss)  ##
#source("00-model-specs.R")

library(xtable)
library(ggplot2)
library(patchwork)
library(broom)

fullfits <- readRDS("fullfits.rds")
cv.auc <- readRDS("cv-aucs-stack.rds")
opt.fit <- readRDS("opt-coeffs.rds")

data("gbsg")

valid <- data.frame(YY = with(gbsg, Surv(rfstime, status)),
                    age = gbsg$age, meno = unclass(gbsg$meno),
                    sizemed = 1.0 * (gbsg$size > 20 & gbsg$size < 50),
                    sizebig = 1.0 * (gbsg$size >50),
                    nodes = unclass(gbsg$nodes), pgr = unclass(gbsg$pgr),
                    er = unclass(gbsg$er), hormon = unclass(gbsg$hormon))
Zvalid <- lapply(fullfits, function(f) f(valid))
YYensvalid <- cumincglm(YY ~ 1, data = valid, time = 5 * 365.25)$y

aucalls <- lapply(Zvalid, function(z) {

  with(calc_roc(z, matrix(YYensvalid, ncol = 1), ramp = smoothramp),
       calc_auc(fpf, tpf))

})

names.models <- c("Cox.stepwise", "survival random forests", "CoxBoost",
                  "IPCW logistic stepwise", "Bagged SVM", "Bagged KNN",
                  "eventglm 2yr", "eventglm 4yr", "eventglm 5yr", "eventglm 6yr", "eventglm 7yr",
                  "LASSO eventglm 2yr", "LASSO eventglm 4yr", "LASSO eventglm 5yr",
                  "LASSO eventglm 6yr", "LASSO eventglm 7yr")


Zmat.valid <- do.call(cbind, Zvalid)

Zfin.opt <- (Zmat.valid %*% (opt.fit$par) / sum(opt.fit$par))[, 1]
roc.opt <- calc_roc(Zfin.opt, matrix(YYensvalid, ncol = 1), ramp = smoothramp)

with(roc.opt, pseudoloss::calc_auc(fpf,tpf))

sumtable <- data.frame(R.package = c("survival", "randomForestSRC", "CoxBoost", "stats",
                                     "e1071", "class", rep("eventglm", 5),
                                     rep("glmnet", 5)),
                       AUC = unlist(aucalls), coefficient = (opt.fit$par) / sum(opt.fit$par))
rownames(sumtable) <- names.models


print(xtable(sumtable[, 1:2], digits = 3))




## roc and predictiveness curve

roc.opt$cut <- sort(unique(Zfin.opt))

p1<- ggplot(roc.opt, aes(x = fpf, y = tpf)) + geom_line() +
  geom_point(data = roc.opt[seq(5, 675, length.out = 6),],
             aes(x = fpf, y = tpf)) +
  geom_text(data = roc.opt[seq(5, 675, length.out = 6),],
            aes(x = fpf - .05, y = tpf + .05, label = round(cut,2))) +
  theme_bw()+ plotROC::style_roc() + geom_abline(intercept = 0, slope = 1, color = "grey90")


Zopt.cut <- pmin(pmax(0, Zfin.opt), .99)


p2 <- ggplot(data.frame(pred.risk = Zopt.cut, pos = YYensvalid), aes(x = pred.risk, y = pos)) +
  stat_smooth(color = "black", method = "loess", span = 1.25, se = FALSE) +
  scale_x_continuous("Predicted 5 year risk") +
  scale_y_continuous("Estimated 5 year risk") +
  geom_rug(aes(x= pred.risk, y = NULL), sides = "b") +
  geom_hline(yintercept = mean(YYensvalid), linetype = 3) +
  coord_cartesian(ylim = c(0, 1)) +
  theme_bw()
p1 + p2
ggsave("int-plot.pdf", width = 7.25, height = 3.75)

valid$Zopt.cut <- Zopt.cut
survConcordance(YY ~ Zopt.cut, data = valid)


valid$group.Z <- cut(valid$Zopt.cut,
                     breaks = c(-1, quantile(valid$Zopt.cut, c(0.25, .5, .75)), 2),
                     labels = c("Q1", "Q2", "Q3", "Q4"))


sfitg <- survfit(YY ~ group.Z, data = valid)
meplot <- tidy(sfitg)
meplot$strata <- gsub("group.Z=", "", meplot$strata, fixed = TRUE)

ggplot(meplot, aes(x = time / 365.25, y = estimate, color = strata)) +
  geom_step() +
  geom_step(aes(y = conf.low), linetype = 3) +
  geom_step(aes(y = conf.high), linetype = 3) +
  theme_classic() +
  scale_color_grey("Risk quartile", start = .7, end = .1) +
  xlab("Years since surgery") + ylab("Survival probability")
ggsave("km-fig.pdf", width = 5.25, height = 3.75)

rpm <- replicate(200, {
  coxfit <- coxph(YY ~ I(Zopt.cut * 10), data = valid[sample(1:nrow(valid), nrow(valid), replace = TRUE), ])
  royston(coxfit)[3]
})
sd(rpm)
summary(coxph(YY ~ I(Zopt.cut * 10), data = valid))
summary(coxph(YY ~ group.Z, data = valid))


aucb <- replicate(200, {

  smi <- sample(1:length(Zfin.opt), length(Zfin.opt), replace = TRUE)
  roc.opt <- calc_roc(Zfin.opt[smi], matrix(YYensvalid[smi], ncol = 1), ramp = smoothramp)
  with(roc.opt, pseudoloss::calc_auc(fpf,tpf))

})
sd(aucb)

print(xtable(data.frame(statistic = c("AUC 5-years",
                                      "Survival concordance",
                                      "Explained variation",
                                      "HR per 0.1 units",
                                      "HR Q2 vs Q1",
                                      "HR Q3 vs Q1",
                                      "HR Q4 vs Q1"),
                        value = c(0.718, 0.685, 0.193, 1.44, 1.96, 3.10, 5.84),
                        se = c(0.022, 0.018, 0.032, 0.033, 0.21, 0.20, 0.19))), include.rownames = FALSE)

